kernel/src/hil/digest.rs

```diff
@@ -214,59 +328,41 @@ pub trait DigestVerify<'a, const L: usize> {
     /// suitable or obvious default option, the implementation can return
     /// `ErrorCode::NOSUPPORT`.
-    fn verify(
-        &'a self,
-        compare: &'static mut [u8; L],
-    ) -> Result<(), (ErrorCode, &'static mut [u8; L])>;
+    fn verify(&'a self, compare: &'static mut D) -> Result<(), (ErrorCode, &'static mut D)>;
 }
 
 /// Computes a digest (cryptographic hash) over data or performs verification.
 ///
-/// 'L' is the length of the 'u8' array to store the digest output.
-pub trait Digest<'a, const L: usize>:
-    DigestData<'a, L> + DigestHash<'a, L> + DigestVerify<'a, L>
+/// `DigestAlgorithm` is the type of digest computed and the data structure to
+/// store the computed digest.
+pub trait Digest<'a, D: DigestAlgorithm>:
+    DigestData<'a, D> + DigestHash<'a, D> + DigestVerify<'a, D>
 {
     /// Set the client instance which will receive `hash_done()`,
     /// `add_data_done()` and `verification_done()` callbacks.
-    fn set_client(&'a self, client: &'a dyn Client<L>);
+    fn set_client(&'a self, client: &'a dyn Client<D>);
 }
 
 /// Computes a digest (cryptographic hash) over data.
 ///
-/// 'L' is the length of the 'u8' array to store the digest output.
-pub trait DigestDataHash<'a, const L: usize>: DigestData<'a, L> + DigestHash<'a, L> {
+/// `DigestAlgorithm` is the type of digest computed and the data structure to
+/// store the computed digest.
+pub trait DigestDataHash<'a, D: DigestAlgorithm>: DigestData<'a, D> + DigestHash<'a, D> {
     /// Set the client instance which will receive `hash_done()` and
     /// `add_data_done()` callbacks.
-    fn set_client(&'a self, client: &'a dyn ClientDataHash<L>);
+    fn set_client(&'a self, client: &'a dyn ClientDataHash<D>);
 }
 
 /// Verify a digest (cryptographic hash) over data.
 ///
-/// 'L' is the length of the 'u8' array to store the digest output.
-pub trait DigestDataVerify<'a, const L: usize>: DigestData<'a, L> + DigestVerify<'a, L> {
+/// `DigestAlgorithm` is the type of digest computed and the data structure to
+/// store the computed digest.
+pub trait DigestDataVerify<'a, D: DigestAlgorithm>:
+    DigestData<'a, D> + DigestVerify<'a, D>
+{
     /// Set the client instance which will receive `verify_done()` and
     /// `add_data_done()` callbacks.
-    fn set_client(&'a self, client: &'a dyn ClientDataVerify<L>);
-}
-
-pub trait Sha224 {
-    /// Call before adding data to perform Sha224
-    fn set_mode_sha224(&self) -> Result<(), ErrorCode>;
-}
-
-pub trait Sha256 {
-    /// Call before adding data to perform Sha256
-    fn set_mode_sha256(&self) -> Result<(), ErrorCode>;
-}
-
-pub trait Sha384 {
-    /// Call before adding data to perform Sha384
-    fn set_mode_sha384(&self) -> Result<(), ErrorCode>;
-}
-
-pub trait Sha512 {
-    /// Call before adding data to perform Sha512
-    fn set_mode_sha512(&self) -> Result<(), ErrorCode>;
+    fn set_client(&'a self, client: &'a dyn ClientDataVerify<D>);
 }
 
 pub trait HmacSha256 {
```

Comment:
Shouldn't these be removed as well?

Reply:
Well, we need to set the key somehow. I'm not sure of a better option, so maybe it's best to just leave them as is?

Reply:
Isn't this confusing? How do you switch between SHA and HMAC modes?